File: examples/go/apps/minipi/scenario_store.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// ScenarioStoreActor — persists eval scenario definitions.
//
// Demonstrates: structured KV storage, scenario catalog, rubric management.
// Each scenario has: input, expected output, rubric type, tags, difficulty.
package main

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// builtinScenarios are seeded at init time.
var builtinScenarios = []map[string]any{
	{
		"scenario_id": "sc-math-01",
		"name":        "Simple multiplication",
		"input":       "What is 6 * 7?",
		"expected":    "42",
		"rubric":      "task_completion",
		"tags":        []any{"math"},
		"difficulty":  "easy",
	},
	{
		"scenario_id": "sc-calc-01",
		"name":        "Step-by-step arithmetic",
		"input":       "Compute (17 * 24) + (89 - 45) step by step",
		"expected":    "452",
		"rubric":      "task_completion",
		"tags":        []any{"math", "calculator"},
		"difficulty":  "easy",
	},
	{
		"scenario_id": "sc-search-01",
		"name":        "Web search intent",
		"input":       "Search for information about the Pythagorean theorem",
		"expected":    nil,
		"rubric":      "tool_use",
		"tags":        []any{"search", "tool_use"},
		"difficulty":  "medium",
	},
	{
		"scenario_id": "sc-reason-01",
		"name":        "Logical deduction",
		"input":       "If all Bloops are Razzies and all Razzies are Lazzies, are all Bloops definitely Lazzies?",
		"expected":    "Yes",
		"rubric":      "task_completion",
		"tags":        []any{"reasoning"},
		"difficulty":  "medium",
	},
	{
		"scenario_id": "sc-budget-01",
		"name":        "Quadratic equation summary",
		"input":       "Summarize the key steps to solve a quadratic equation ax^2 + bx + c = 0",
		"expected":    nil,
		"rubric":      "task_completion",
		"tags":        []any{"math", "summary"},
		"difficulty":  "medium",
	},
	{
		"scenario_id": "sc-contract-01",
		"name":        "Expression validation",
		"input":       "Validate: is the expression '(2 + 3) * (4 - 1)' valid? What is its value?",
		"expected":    "15",
		"rubric":      "task_completion",
		"tags":        []any{"validation", "math"},
		"difficulty":  "easy",
	},
	{
		"scenario_id": "sc-multi-01",
		"name":        "Multi-step tool use",
		"input":       "Search for the capital of France, then compute 3 * 7, then report both results",
		"expected":    nil,
		"rubric":      "tool_use",
		"tags":        []any{"multi-step", "search", "tool_use"},
		"difficulty":  "hard",
	},
	{
		"scenario_id": "sc-kv-01",
		"name":        "KV store round-trip",
		"input":       "Store the value 'hello world' under key 'test_key', then read it back and verify",
		"expected":    nil,
		"rubric":      "tool_use",
		"tags":        []any{"kv", "tool_use"},
		"difficulty":  "medium",
	},
	{
		"scenario_id": "sc-chain-01",
		"name":        "Chained computation",
		"input":       "Compute sqrt(144), then add 5 to the result, then multiply by 2",
		"expected":    "34",
		"rubric":      "task_completion",
		"tags":        []any{"math", "chain"},
		"difficulty":  "medium",
	},
	{
		"scenario_id": "sc-compare-01",
		"name":        "Power comparison",
		"input":       "Which is larger: 2^10 or 10^3? Show your calculation",
		"expected":    "2^10",
		"rubric":      "task_completion",
		"tags":        []any{"math", "comparison"},
		"difficulty":  "easy",
	},
}

// ScenarioStoreActor stores, retrieves, and lists eval scenarios.
type ScenarioStoreActor struct {
	plexspaces.BaseActor
	ActorID       string `json:"actor_id"`
	ScenarioCount int    `json:"scenario_count"`
}

func NewScenarioStoreActor() plexspaces.Actor {
	a := &ScenarioStoreActor{}
	a.SetSelf(a)
	return a
}

func (s *ScenarioStoreActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	s.SetRuntimeMetadata(config.ActorID)
	s.ActorID = config.ActorID
	if err := host.PG().Join("svc:scenario_store"); err != nil {
		host.Warn(fmt.Sprintf("ScenarioStoreActor: failed to join svc:scenario_store: %v", err))
	}
	s.seedBuiltinScenarios()
	host.Info(fmt.Sprintf("ScenarioStoreActor Init actor_id=%s", config.ActorID))
	return ""
}

func (s *ScenarioStoreActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "get_scenario":
		return s.getScenario(p)
	case "list_scenarios":
		return s.listScenarios(p)
	case "put_scenario":
		return s.putScenario(p)
	case "get_suite":
		return s.getSuite(p)
	case "put_suite":
		return s.putSuite(p)
	case "get_stats":
		return marshal(map[string]any{
			"status":         "ok",
			"actor_id":       s.ActorID,
			"scenario_count": s.ScenarioCount,
		})
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (s *ScenarioStoreActor) getScenario(p map[string]any) string {
	scenarioID := stringVal(p, "scenario_id", "")
	if scenarioID == "" {
		return marshal(map[string]any{"error": "scenario_id is required"})
	}
	raw := host.KVGet("scenario:" + scenarioID)
	if raw == "" {
		return marshal(map[string]any{"error": fmt.Sprintf("scenario %s not found", scenarioID)})
	}
	var sc map[string]any
	if err := json.Unmarshal([]byte(raw), &sc); err != nil {
		return marshal(map[string]any{"error": "failed to parse scenario"})
	}
	return marshal(map[string]any{"status": "ok", "scenario": sc})
}

func (s *ScenarioStoreActor) listScenarios(p map[string]any) string {
	difficulty := stringVal(p, "difficulty", "")
	limit := intVal(p, "limit", 50)
	tagsRaw := sliceVal(p, "tags")
	tags := extractStringSliceFromAny(tagsRaw)

	// Load all known scenario IDs from our seeded scenarios
	scenarios := make([]any, 0)
	for _, sc := range builtinScenarios {
		if difficulty != "" && sc["difficulty"] != difficulty {
			continue
		}
		if len(tags) > 0 {
			scTags := extractStringSliceFromAny(sc["tags"].([]any))
			matched := false
			for _, t := range tags {
				for _, st := range scTags {
					if t == st {
						matched = true
						break
					}
				}
				if matched {
					break
				}
			}
			if !matched {
				continue
			}
		}
		scID := sc["scenario_id"].(string)
		raw := host.KVGet("scenario:" + scID)
		if raw != "" {
			var loaded map[string]any
			if err := json.Unmarshal([]byte(raw), &loaded); err == nil {
				scenarios = append(scenarios, loaded)
			}
		} else {
			scenarios = append(scenarios, sc)
		}
		if len(scenarios) >= limit {
			break
		}
	}
	return marshal(map[string]any{"status": "ok", "scenarios": scenarios, "count": len(scenarios)})
}

func (s *ScenarioStoreActor) putScenario(p map[string]any) string {
	scenario, _ := p["scenario"].(map[string]any)
	if scenario == nil {
		return marshal(map[string]any{"error": "scenario is required"})
	}
	scenarioID := stringVal(scenario, "scenario_id", "")
	if scenarioID == "" {
		scenarioID = fmt.Sprintf("sc-%d", host.NowMs())
		scenario["scenario_id"] = scenarioID
	}
	scJSON, _ := json.Marshal(scenario)
	host.KVPut("scenario:"+scenarioID, string(scJSON))
	s.ScenarioCount++
	s.IncrCounter(host, "scenarios_stored_total")
	return marshal(map[string]any{"status": "ok", "scenario_id": scenarioID})
}

func (s *ScenarioStoreActor) getSuite(p map[string]any) string {
	suiteName := stringVal(p, "suite_name", "")
	scenarioIDsRaw := sliceVal(p, "scenario_ids")
	scenarioIDs := extractStringSliceFromAny(scenarioIDsRaw)

	if len(scenarioIDs) > 0 {
		scenarios := s.loadScenariosByIDs(scenarioIDs)
		return marshal(map[string]any{"status": "ok", "suite_name": suiteName, "scenarios": scenarios, "count": len(scenarios)})
	}

	// Named suites
	var ids []string
	switch suiteName {
	case "smoke":
		ids = []string{"sc-math-01"}
	case "standard":
		ids = []string{"sc-math-01", "sc-calc-01", "sc-search-01", "sc-reason-01", "sc-budget-01"}
	case "full":
		for _, sc := range builtinScenarios {
			ids = append(ids, sc["scenario_id"].(string))
		}
	default:
		// Try stored suite definition
		raw := host.KVGet("suite:" + suiteName)
		if raw == "" {
			return marshal(map[string]any{"error": fmt.Sprintf("unknown suite: %s", suiteName)})
		}
		var suiteDef map[string]any
		if err := json.Unmarshal([]byte(raw), &suiteDef); err != nil {
			return marshal(map[string]any{"error": "failed to parse suite"})
		}
		idsRaw := sliceVal(suiteDef, "scenario_ids")
		ids = extractStringSliceFromAny(idsRaw)
	}

	scenarios := s.loadScenariosByIDs(ids)
	return marshal(map[string]any{"status": "ok", "suite_name": suiteName, "scenarios": scenarios, "count": len(scenarios)})
}

func (s *ScenarioStoreActor) loadScenariosByIDs(ids []string) []any {
	scenarios := make([]any, 0, len(ids))
	for _, id := range ids {
		raw := host.KVGet("scenario:" + id)
		if raw != "" {
			var sc map[string]any
			if err := json.Unmarshal([]byte(raw), &sc); err == nil {
				scenarios = append(scenarios, sc)
				continue
			}
		}
		// Fall back to builtin if not in KV
		for _, sc := range builtinScenarios {
			if sc["scenario_id"] == id {
				scenarios = append(scenarios, sc)
				break
			}
		}
	}
	return scenarios
}

func (s *ScenarioStoreActor) putSuite(p map[string]any) string {
	suiteName := stringVal(p, "suite_name", "")
	scenarioIDsRaw := sliceVal(p, "scenario_ids")
	if suiteName == "" || len(scenarioIDsRaw) == 0 {
		return marshal(map[string]any{"error": "suite_name and scenario_ids are required"})
	}
	ids := extractStringSliceFromAny(scenarioIDsRaw)
	suiteJSON, _ := json.Marshal(map[string]any{"scenario_ids": ids})
	host.KVPut("suite:"+suiteName, string(suiteJSON))
	return marshal(map[string]any{"status": "ok", "suite_name": suiteName, "count": len(ids)})
}

func (s *ScenarioStoreActor) seedBuiltinScenarios() {
	seeded := 0
	for _, sc := range builtinScenarios {
		id := sc["scenario_id"].(string)
		if existing := host.KVGet("scenario:" + id); existing == "" {
			scJSON, _ := json.Marshal(sc)
			host.KVPut("scenario:"+id, string(scJSON))
			seeded++
		}
	}
	s.ScenarioCount = len(builtinScenarios)
	if seeded > 0 {
		host.Info(fmt.Sprintf("ScenarioStoreActor seeded %d built-in scenarios (ids: %s)",
			seeded, strings.Join(func() []string {
				ids := make([]string, 0, len(builtinScenarios))
				for _, sc := range builtinScenarios {
					ids = append(ids, sc["scenario_id"].(string))
				}
				return ids
			}(), ", ")))
	}
}

File: examples/python/apps/minipi/scenario_store.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""ScenarioStoreActor — persists eval scenario definitions.

Demonstrates: structured KV storage, scenario catalog, rubric management.
Each scenario has: input, expected output, rubric type, tags, difficulty.
"""

import json
from plexspaces import actor, state, init_handler, handler, host

# Built-in scenarios seeded at init time
_BUILTIN_SCENARIOS = [
    {
        "scenario_id": "sc-math-01",
        "name": "Simple multiplication",
        "input": "What is 6 * 7?",
        "expected": "42",
        "rubric": "task_completion",
        "tags": ["math"],
        "difficulty": "easy",
    },
    {
        "scenario_id": "sc-calc-01",
        "name": "Step-by-step arithmetic",
        "input": "Compute (17 * 24) + (89 - 45) step by step",
        "expected": "452",
        "rubric": "task_completion",
        "tags": ["math", "calculator"],
        "difficulty": "easy",
    },
    {
        "scenario_id": "sc-search-01",
        "name": "Web search intent",
        "input": "Search for information about the Pythagorean theorem",
        "expected": None,  # Open-ended — scored by rubric
        "rubric": "tool_use",
        "tags": ["search", "tool_use"],
        "difficulty": "medium",
    },
    {
        "scenario_id": "sc-reason-01",
        "name": "Logical deduction",
        "input": "If all Bloops are Razzies and all Razzies are Lazzies, are all Bloops definitely Lazzies?",
        "expected": "Yes",
        "rubric": "task_completion",
        "tags": ["reasoning"],
        "difficulty": "medium",
    },
    {
        "scenario_id": "sc-budget-01",
        "name": "Quadratic equation summary",
        "input": "Summarize the key steps to solve a quadratic equation ax^2 + bx + c = 0",
        "expected": None,
        "rubric": "task_completion",
        "tags": ["math", "summary"],
        "difficulty": "medium",
    },
    {
        "scenario_id": "sc-contract-01",
        "name": "Expression validation",
        "input": "Validate: is the expression '(2 + 3) * (4 - 1)' valid? What is its value?",
        "expected": "15",
        "rubric": "task_completion",
        "tags": ["validation", "math"],
        "difficulty": "easy",
    },
    {
        "scenario_id": "sc-multi-01",
        "name": "Multi-step tool use",
        "input": "Search for the capital of France, then compute 3 * 7, then report both results",
        "expected": None,
        "rubric": "tool_use",
        "tags": ["multi-step", "search", "tool_use"],
        "difficulty": "hard",
    },
    {
        "scenario_id": "sc-kv-01",
        "name": "KV store round-trip",
        "input": "Store the value 'hello world' under key 'test_key', then read it back and verify",
        "expected": None,
        "rubric": "tool_use",
        "tags": ["kv", "tool_use"],
        "difficulty": "medium",
    },
    {
        "scenario_id": "sc-chain-01",
        "name": "Chained computation",
        "input": "Compute sqrt(144), then add 5 to the result, then multiply by 2",
        "expected": "34",
        "rubric": "task_completion",
        "tags": ["math", "chain"],
        "difficulty": "medium",
    },
    {
        "scenario_id": "sc-compare-01",
        "name": "Power comparison",
        "input": "Which is larger: 2^10 or 10^3? Show your calculation",
        "expected": "2^10",
        "rubric": "task_completion",
        "tags": ["math", "comparison"],
        "difficulty": "easy",
    },
]


@actor
class ScenarioStoreActor:
    """
    Scenario catalog: stores, retrieves, and lists eval scenarios.

    Scenarios are persisted in KV storage keyed by scenario_id.
    Built-in scenarios are seeded at init time.

    A scenario contains:
    - input: what to tell the agent
    - expected: expected output (None for open-ended)
    - rubric: scoring method (task_completion, tool_use, efficiency, llm_judge)
    - tags: for filtering
    - difficulty: easy/medium/hard
    """

    actor_id: str = state(default="")
    scenario_count: int = state(default=0)

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        try:
            host.kv.put("svc:scenario_store", host.self_id())
        except Exception:
            pass
        try:
            host.registry.register(None, self.actor_id or host.self_id(), "actor", "",
                                   object_category="scenario_store")
        except Exception:
            pass
        host.info(f"ScenarioStoreActor init actor_id={self.actor_id}")
        self._seed_builtin_scenarios()

    @handler("get_scenario")
    def get_scenario(self, scenario_id: str = "") -> dict:
        """Get a single scenario by ID."""
        if not scenario_id:
            return {"error": "scenario_id is required"}
        raw = host.kv.get(f"scenario:{scenario_id}")
        if not raw:
            return {"error": f"scenario {scenario_id} not found"}
        try:
            return {"status": "ok", "scenario": json.loads(raw)}
        except Exception:
            return {"error": "failed to parse scenario"}

    @handler("list_scenarios")
    def list_scenarios(self, tags: list = None, difficulty: str = "", limit: int = 50) -> dict:
        """List scenarios, optionally filtered by tags or difficulty."""
        try:
            raw_keys = host.kv.list("scenario:")
            keys = json.loads(raw_keys) if raw_keys and not raw_keys.startswith("ERROR:") else []
            scenarios = []
            for key in keys[:limit * 2]:  # Oversample to account for filtering
                raw = host.kv.get(key)
                if not raw:
                    continue
                try:
                    sc = json.loads(raw)
                except Exception:
                    continue

                if difficulty and sc.get("difficulty") != difficulty:
                    continue
                if tags:
                    sc_tags = sc.get("tags", [])
                    if not any(t in sc_tags for t in tags):
                        continue

                scenarios.append(sc)
                if len(scenarios) >= limit:
                    break

            return {"status": "ok", "scenarios": scenarios, "count": len(scenarios)}
        except Exception as e:
            return {"error": str(e)}

    @handler("put_scenario")
    def put_scenario(self, scenario: dict = None) -> dict:
        """Store or update a scenario."""
        if not scenario:
            return {"error": "scenario is required"}
        scenario_id = scenario.get("scenario_id", "")
        if not scenario_id:
            scenario_id = host.new_id()
            scenario["scenario_id"] = scenario_id

        try:
            host.kv.put(f"scenario:{scenario_id}", json.dumps(scenario))
            self.scenario_count += 1
            host.incr_counter("scenarios_stored_total", 1)
            return {"status": "ok", "scenario_id": scenario_id}
        except Exception as e:
            return {"error": str(e)}

    @handler("get_suite")
    def get_suite(self, suite_name: str = "", scenario_ids: list = None) -> dict:
        """
        Get a named suite of scenarios.

        Built-in suites:
        - "smoke": 1 easy scenario
        - "standard": all easy + medium scenarios
        - "full": all scenarios

        Or pass explicit scenario_ids.
        """
        if scenario_ids:
            scenarios = []
            for sid in scenario_ids:
                raw = host.kv.get(f"scenario:{sid}")
                if raw:
                    try:
                        scenarios.append(json.loads(raw))
                    except Exception:
                        pass
            return {"status": "ok", "suite_name": suite_name, "scenarios": scenarios}

        # Named suites
        if suite_name == "smoke":
            ids = ["sc-math-01"]
        elif suite_name == "standard":
            ids = ["sc-math-01", "sc-calc-01", "sc-search-01", "sc-reason-01", "sc-budget-01"]
        elif suite_name == "full":
            ids = [s["scenario_id"] for s in _BUILTIN_SCENARIOS]
        else:
            # Try to load a stored suite definition
            raw = host.kv.get(f"suite:{suite_name}")
            if raw:
                try:
                    suite_def = json.loads(raw)
                    ids = suite_def.get("scenario_ids", [])
                except Exception:
                    ids = []
            else:
                return {"error": f"unknown suite: {suite_name}"}

        scenarios = []
        for sid in ids:
            raw = host.kv.get(f"scenario:{sid}")
            if raw:
                try:
                    scenarios.append(json.loads(raw))
                except Exception:
                    pass

        return {"status": "ok", "suite_name": suite_name, "scenarios": scenarios, "count": len(scenarios)}

    @handler("put_suite")
    def put_suite(self, suite_name: str = "", scenario_ids: list = None) -> dict:
        """Define a named suite."""
        if not suite_name or not scenario_ids:
            return {"error": "suite_name and scenario_ids are required"}
        try:
            host.kv.put(f"suite:{suite_name}", json.dumps({"scenario_ids": scenario_ids}))
            return {"status": "ok", "suite_name": suite_name, "count": len(scenario_ids)}
        except Exception as e:
            return {"error": str(e)}

    @handler("get_stats")
    def get_stats(self) -> dict:
        return {
            "status": "ok",
            "actor_id": self.actor_id,
            "scenario_count": self.scenario_count,
        }

    # ------------------------------------------------------------------

    def _seed_builtin_scenarios(self) -> None:
        """Seed built-in scenarios if they don't already exist."""
        seeded = 0
        for sc in _BUILTIN_SCENARIOS:
            key = f"scenario:{sc['scenario_id']}"
            existing = host.kv.get(key)
            if not existing:
                try:
                    host.kv.put(key, json.dumps(sc))
                    seeded += 1
                except Exception as e:
                    host.warn(f"Failed to seed scenario {sc['scenario_id']}: {e}")

        self.scenario_count = len(_BUILTIN_SCENARIOS)
        if seeded > 0:
            host.info(f"ScenarioStoreActor seeded {seeded} built-in scenarios")
